[10.x.x] Add SpeedTree8 shadergraph to HDRP and URP

nerites opened the pull request:
# **Please read the [Contributing guide](CONTRIBUTING.md) before making a PR.**

* Read the [Graphics repository & Yamato FAQ](http://go/graphics-yamato-faq).

### Checklist for PR maker
- [x] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [x] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [x] Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR. If you do add documentation, make sure to add the relevant Graphics Docs team member as a reviewer of the PR. If you are not sure which person to add, see the [Docs team contacts sheet](https://docs.google.com/spreadsheets/d/1rgUWWgwLFEHIQ3Rz-LnK6PAKmbM49DZZ9al4hvnztOo/edit#gid=1058860420).
- [x] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
Adding SpeedTree8 support (including material upgrader) to HDRP.
Added the SpeedTree8 shadergraph to URP as well, though the default is still the original SpeedTree8 shader pending wind support in the shadergraph.
No changes from the 7.x.x version of this PR, aside from what was necessary to merge with 10.x.x version of HDRP.

21.1 PR: https://github.com/Unity-Technologies/Graphics/pull/4068
PR to master: https://github.com/Unity-Technologies/Graphics/pull/3911/

---
### Testing status
(Copied from 7.x.x)
Tested importing and material upgrading in HDRP. It works, except for an issue with embedded materials where when the Library is nuked, the materials don't get setup properly. This is because the speedtree asset is imported before the HDRP asset is imported, so the correct shader is not grabbed, and the correct postprocessing is not run. This is a recurring issue and topic of discussion not unique to SpeedTree, so it should not be blocking for this PR.

For URP, tested that manually switching over the shader works as expected. No upgrader or importer support for the shadergraph is needed yet as it is not yet the default shader.

---
### Comments to reviewers
@adrien-de-tocqueville since you signed off on the 7.x.x version of this, and the 10.x.x is unchanged.
@LandonTownsendUnity for a QA pass on HDRP (should behave the same as 7.x.x). 

Shader graph: Please confirm placement of new subshadergraphs I've added to the shadergraph package, including changelog notes. I also added a LODCrossfade hlsl file. This is not ST8-specific; is there a better way of adding this?

adrien-de-tocqueville commented:
Sorry about the missclick ! Marked the PR as ready by mistake, but it looks like I cannot remove the reviewers that were automatically added

adrien-de-tocqueville reviewed (approved):
Looks the same as the PR for 7.x.x

elizabeth-legros reviewed (approved):
Pending docs for lodcrossfade looks good on my end

ValGrimm-U3D commented on com.unity.render-pipelines.core/Editor/SpeedTree8MaterialUpgrader.cs:
```diff
@@ -0,0 +1,181 @@
+        };
+
+        /// <summary>
+        /// Creates a material upgrader with only the renames in common between HD and Universal.
```
Commenting on this as it appears to be API doc.
Not clear what "only" means here. Does it mean that renaming materials is the only thing the material upgrader does the same way in both HDRP and URP?

nerites replied:
If there's a venn diagram of renames that need to happen, it goes like this:
(properties only in HD ( properties in both ) properties only in universal)
"only" here refers to the intersection of the venn diagram. There's probably a better way of writing this though

nerites replied:
Changed to:
"Creates a material upgrader that handles the property renames that HD and Universal have in common."

nerites replied:
Changed again, and integrated all your feedback:

        /// <summary>
        /// Creates a material upgrader that handles the property renames that HD and Universal have in common when upgrading
        /// from the built-in SpeedTree8 shader.
        /// </summary>
        /// <param name="sourceShaderName">Original SpeedTree8 shader name.</param>
        /// <param name="destShaderName">New SpeedTree8 shader name.</param>
        /// <param name="finalizer">A delegate that postprocesses the material for the render pipeline in use.</param>

ValGrimm-U3D replied:
Looks good to me. Thanks!

ValGrimm-U3D commented on com.unity.render-pipelines.core/Editor/SpeedTree8MaterialUpgrader.cs:
```diff
@@ -0,0 +1,181 @@
+        /// <summary>
+        /// Creates a material upgrader with only the renames in common between HD and Universal.
+        /// </summary>
+        /// <param name="sourceShaderName">Original ST8 shader name.</param>
```
Will the user know the ST8 abbreviation?

nerites replied:
Replaced all references to ST8/SpeedTree8 with SpeedTree 8

ValGrimm-U3D replied:
Great!

ValGrimm-U3D commented on com.unity.render-pipelines.core/Editor/SpeedTree8MaterialUpgrader.cs:
```diff
@@ -0,0 +1,181 @@
+        /// </summary>
+        /// <param name="sourceShaderName">Original ST8 shader name.</param>
+        /// <param name="destShaderName">New ST8 shader name.</param>
+        /// <param name="finalizer">A delegate that postprocesses the material as needed by the render pipeline.</param>
```
"as needed" is a bit imprecise. 
Would this rewrite be correct?
`A delegate that postprocesses the material for the render pipeline in use.`

ValGrimm-U3D commented on com.unity.render-pipelines.core/Editor/SpeedTree8MaterialUpgrader.cs:
```diff
@@ -0,0 +1,181 @@
+        }
+
+        /// <summary>
+        /// Postprocess materials when importing a SpeedTree8 asset. Call from OnPostprocessSpeedTree in a MaterialPostprocessor.
```
The vendor seems to prefer SpeedTree 8 to SpeedTree8. Also potentially confusing because this document refers to ST8 in at least one other summary.

ValGrimm-U3D replied:
Postprocesses materials while you are importing a SpeedTree 8 asset. Call from OnPostprocessSpeedTree in a MaterialPostprocessor.

(need to make sure that user knows how to call things from a MaterialPostprocessor. A link  or more info is likely needed).

nerites replied:
How do you add links in these comment-style docs? I think a link to MaterialPostprocessor documentation would be sufficient here.

That documentation might not exist, but it shouldn't be added in this PR.

ValGrimm-U3D replied:
Not sure we can link, actually, and didn't want to add full doc for that to this PR, just a "see the doc for" kind of thing. Don't seem to find that doc, so I'll create a ticket. What about the other proposed changes?

nerites replied:
I copy-pasted the rewrite you proposed in c29ee75

ValGrimm-U3D commented on com.unity.render-pipelines.core/Editor/SpeedTree8MaterialUpgrader.cs:
```diff
@@ -0,0 +1,181 @@
+        /// <summary>
+        /// Postprocess materials when importing a SpeedTree8 asset. Call from OnPostprocessSpeedTree in a MaterialPostprocessor.
+        /// </summary>
+        /// <param name="speedtree">Game object for the SpeedTree asset being imported.</param>
```
"for" is unclear here. 
Would this rewrite be correct?
`The GameObject Unity creates from this imported SpeedTree.`

ValGrimm-U3D commented on com.unity.render-pipelines.core/Editor/SpeedTree8MaterialUpgrader.cs:
```diff
@@ -0,0 +1,181 @@
+        /// Postprocess materials when importing a SpeedTree8 asset. Call from OnPostprocessSpeedTree in a MaterialPostprocessor.
+        /// </summary>
+        /// <param name="speedtree">Game object for the SpeedTree asset being imported.</param>
+        /// <param name="stImporter">The assetimporter used to import the SpeedTree asset.</param>
```
The asset importer Unity uses to import this SpeedTree asset.

ValGrimm-U3D commented on com.unity.render-pipelines.high-definition/Documentation~/Feature-Comparison.md:
```diff
@@ -306,7 +306,7 @@ Managing LOD in this way has two consequences:
 | **Feature**                  | **Built-in Render Pipeline** | **High Definition Render Pipeline (HDRP)** |
 | ------------------------ | ------------------------ | ------------------------------- |
-| **Speed Tree**           | Yes                      | In research                     |
+| **Speed Tree**           | Yes                      | SpeedTree8 only. Expect visual changes for correctness in future versions. |
```
Does "future versions" refer to versions of HDRP, Unity, or SpeedTree?
Also, should be SpeedTree 8 with a space, according to the vendor website.

nerites replied:
Future versions of HD and Universal SpeedTree 8 shadergraphs. It's referring specifically to the calculation of normals using custom interpolators. It's kind of behind-the-scenes... do you think it's worth mentioning? This comment should be removed in the 21.2+ PR.

ValGrimm-U3D replied:
Ok, based on what you've said I've revised that to:

Currently HDRP and URP only support SpeedTree 8. In a future version of Shader Graph, this functionality will be visually different because of changes to normal calculation with custom interpolators.

nerites replied:
This is the HD feature comparison page. Does it make sense to mention universal here?

In c29ee75 I wrote it as the following, what do you think?
SpeedTree 8 only. Expect visual changes to normals for correctness in 21.2.

ValGrimm-U3D replied:
Oh right, it is the HDRP page. Although I thought the visual changes were coming in 22.1?

ValGrimm-U3D replied:
> SpeedTree 8 only. Expect visual changes to normals for correctness in 21.2.
is fine, but my sense was that the visual change was not in scope for 21.2?

nerites replied:
Oh I see-- when I said 21.2+, I meant 21.2 and later will have the updated normals.

ValGrimm-U3D replied:
I am a bit confused. 

What I understand is:
This feature is being introduced in 21.2.
The normals are being changed in 21.2.
Why then state that the normals will change when they are part of the same release?

I also thought the delay in this change to the normals was why this shader won't be the default in URP until 22.1?

ValGrimm-U3D commented on com.unity.render-pipelines.core/Editor/SpeedTree8MaterialUpgrader.cs:
```diff
@@ -0,0 +1,181 @@
+
+        /// <summary>
+        /// Preserve wind quality and billboard setting when upgrading a ST8 material from previous versions.
+        /// Wind priority order is enabled keyword > _WindQuality float value.
```
What is "wind priority"? 
I think the sentence would go something like: "To enable wind priority, use the _WindQuality keyword . . . "
but it isn't clear how the float value relates to this.

ValGrimm-U3D replied:
This code looks like it could be related, defining the _WindQuality keyword . . . 

`windQuality = material.HasProperty("_WindQuality") ? (int)material.GetFloat("_WindQuality") : 0;`

nerites replied:
Wind priority here refers to what the code prioritizes as the "authority" when guessing what WindQuality to set to a material. 

First it checks the keywords- if a quality is set there, it's the source of truth. Otherwise, we check the value set to the float property. 

This is only relevant because HD does some keyword resetting when synchronizing its materials.

ValGrimm-U3D replied:
Thanks!
So how does the float value determine the WindQuality? If the float is higher presumably it has a specific effect, and if it is lower presumably it has a different effect.

Progressing in the rewrite, but don't yet have all the info . . . 
To determine which WindQuality to use, Unity checks the keywords first and then the _WindQuality float value. [explain about float value here].

nerites replied:
Wind Quality is defined by speed tree. Speed Tree has documentation here: 
https://docs8.speedtree.com/modeler/doku.php?id=windunity

ValGrimm-U3D replied:
Oh right, not Unity-specific. That said, the SpeedTree doc doesn't clarify which float values correspond with which quality levels and I can't seem to find that info anywhere. It is important that folks know what the float value means.

nerites replied:
So we should probably add it to the main SpeedTree documentation page in Unity: 
https://docs.unity3d.com/Manual/SpeedTree.html

The quality settings are:
0 = None
1 = Fastest
2 = Faster
3 = Better
4 = Best
5 = Palm

nerites replied:
Does the comments for this function need any change though?

ValGrimm-U3D replied:
Ok, I'd propose that I arrange to update the main SpeedTree doc page to add then quality settings.m
With that in mind:

To determine which WindQuality to use, Unity checks the keywords first and then the _WindQuality float value. See SpeedTree in the Unity Manual for the values associated with different WindQuality settings.

ValGrimm-U3D replied:
Added a note to the ticket about updating manual page for speedtree as well, I will do it if not owned by Sharlene.

ValGrimm-U3D commented on com.unity.render-pipelines.core/Editor/SpeedTree8MaterialUpgrader.cs:
```diff
@@ -0,0 +1,181 @@
+        }
+
+        /// <summary>
+        /// Preserve wind quality and billboard setting when upgrading a ST8 material from previous versions.
```
Proposed rewrite:
Preserves wind quality and billboard settings while you are upgrading a SpeedTree 8 material from previous versions of SpeedTree.

nerites replied:
Slight modification:
"Preserves wind quality and billboard settings while you are upgrading a SpeedTree 8 material from previous versions of SpeedTree 8."

Upgrading from SpeedTree 7 to SpeedTree 8, for example, is not being supported here.

ValGrimm-U3D replied:
Great. Thanks for the clarification!

ValGrimm-U3D commented on com.unity.render-pipelines.core/Editor/SpeedTree8MaterialUpgrader.cs:
```diff
@@ -0,0 +1,181 @@
+
+        private static int GetWindQuality(Material material, int windQuality = -1)
+        {
+            // Conservative wind quality priority:
```
Does quality priority relate to a failover from higher to lower quality?
Does "conservative" mean more likely to failover, or less likely to failover? (Or something else, if this doesn't have to do with failover)

ValGrimm-U3D commented on com.unity.render-pipelines.core/Editor/SpeedTree8MaterialUpgrader.cs:
```diff
@@ -0,0 +1,181 @@
+        /// Wind priority order is enabled keyword > _WindQuality float value.
+        /// Should work for upgrading versions within a pipeline and from standard to current pipeline.
+        /// </summary>
+        /// <param name="material">SpeedTree8 material to upgrade.</param>
```
SpeedTree 8

ValGrimm-U3D commented on com.unity.render-pipelines.core/Editor/SpeedTree8MaterialUpgrader.cs:
```diff
@@ -0,0 +1,181 @@
+        private static int GetWindQuality(Material material, int windQuality = -1)
+        {
+            // Conservative wind quality priority:
+            // input WindQuality > enabled keyword > _WindQuality float value
```
This is a bit cryptic. What would the user be trying to do with the keyword here?

ValGrimm-U3D commented on com.unity.render-pipelines.high-definition/Editor/Material/Nature/HDSpeedTree8MaterialUpgrader.cs:
```diff
@@ -0,0 +1,104 @@
+namespace UnityEditor.Rendering.HighDefinition
+{
+    /// <summary>
+    /// SpeedTree8 material upgrader for HDRP.
```
SpeedTree 8

ValGrimm-U3D commented on com.unity.render-pipelines.high-definition/Editor/Material/Nature/HDSpeedTree8MaterialUpgrader.cs:
```diff
@@ -0,0 +1,104 @@
+        /// Creates a SpeedTree8 material upgrader for HDRP.
+        /// </summary>
+        /// <param name="sourceShaderName">Original shader name.</param>
+        /// <param name="destShaderName">Upgrade shader name.</param>
```
Upgraded shader name

ValGrimm-U3D commented on com.unity.render-pipelines.high-definition/Editor/Material/Nature/HDSpeedTree8MaterialUpgrader.cs:
```diff
@@ -0,0 +1,104 @@
+            HDShaderUtils.ResetMaterialKeywords(mat);
+        }
+        /// <summary>
+        /// Checks if a given material is an HD SpeedTree8 material.
```
proposed rewrite:
Determines if a given material is associated with SpeedTree 8 and compatible with HDRP.

nerites replied:
What about:
"Determines if a given material is using the default SpeedTree 8 shader for HDRP."

ValGrimm-U3D replied:
Looks good to me.

ValGrimm-U3D commented on com.unity.render-pipelines.high-definition/Editor/Material/Nature/HDSpeedTree8MaterialUpgrader.cs:
```diff
@@ -0,0 +1,104 @@
+        /// <summary>
+        /// Checks if a given material is an HD SpeedTree8 material.
+        /// </summary>
+        /// <param name="mat">Material to check.</param>
```
Proposed rewrite:
Material to check for compatibility with SpeedTree and HDRP

nerites replied:
What about:
"Material to check for an HDRP-compatible SpeedTree 8 shader."

ValGrimm-U3D replied:
Looks good to me!

ValGrimm-U3D commented on com.unity.render-pipelines.high-definition/Editor/Material/Nature/HDSpeedTree8MaterialUpgrader.cs:
```diff
@@ -0,0 +1,104 @@
+        }
+
+        /// <summary>
+        /// Restores SpeedTree8-specific material properties and keywords that were set during import and should not be reset.
```
Proposed rerwrite:
During import, the Editor may modify SpeedTree-specific material properties and keywords which should not be modified. This keyword restores these properties to their original state.

nerites replied:
What about:
"During normal usage, HDRP may reset SpeedTree-specific keywords which should not be modified. This method restores these properties to their original state."

ValGrimm-U3D replied:
What is "normal usage"?

nerites replied:
When you click on a material for the first time after it's created, it resets that material's keywords. Comments say this is a hack and should probably not be mentioned specifically. 
Importing and upgrading a material also triggers a reset of that material keyword.

ValGrimm-U3D replied:
"HDRP may reset SpeedTree-specific keywords which should not be modified. This method restores these keywords to their original state."

I propose leaving out the part about "normal usage" and just say it may reset things, because otherwise folks may want to dig into when and why.

nerites replied:
Sounds great, thanks!

ValGrimm-U3D commented on com.unity.render-pipelines.core/Editor/SpeedTree8MaterialUpgrader.cs:
```diff
@@ -0,0 +1,181 @@
+namespace UnityEditor.Rendering
+{
+	/// <summary>
+	/// Material upgrader and relevant utilities for SpeedTree8.
```
SpeedTree 8

ValGrimm-U3D commented on com.unity.shadergraph/CHANGELOG.md:
```diff
@@ -14,6 +14,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ### Added
 - Added padding to the shader graph blackboard window to prevent scrollbars from being overlapped by the resize region at the boundaries of the window
 - Support for the XboxSeries platform has been added.
+- Added subshadergraphs for SpeedTree8 shadergraph support: SpeedTree8Wind, SpeedTree8ColorAlpha, SpeedTree8Billboard. Includes a LODCrossfade hlsl that can be used by anything with a LODGroup.
```
"Includes a LODCrossfade hlsl that can be used"

A "a LODCrossfade hlsl" what--a script?

nerites replied:
It's an HLSL shader snippet. It's intended to be included in either handwritten shaders or via a custom function node in Shadergraph.

ValGrimm-U3D replied:
Ok, that's helpful. How is this rewrite?

Includes a HLSL snippet which enables you to use LODCrossfade functionality with any GameObject in an LODGroup.

nerites replied:
A LODGroup is a component on a prefab or prefab instance. Each LOD contains its own mesh and materials. I guess a prefab instance is technically a type of game object? But it has a LODGroup, not the other way around. 

How about: "Includes an hlsl file that can be included to crossfade between the LODs of any prefab with a LODGroup component."

nerites replied:
On second thought, I'd like to break the LODCrossfade info into its own line as follows:
"Adds an HLSL file that implements a ShaderGraph-compatible version of the LODDitheringTransition function defined in core."

The functionality in LODCrossfade exists in Common.hlsl in core, however that function uses an HLSL intrinsic called clip(), and doesn't output anything that can be hooked into a shadergraph as a dependency. Because of this, that code gets optimized away and never runs. So this version of LODCrossfade actually is shadergraph specific. Maybe it should be renamed to reflect all this though. Maybe to LODDitheringTransitionSG?

ValGrimm-U3D replied:
Renaming the function sounds like a very good idea. If you do not rename it, this would be my propose rerwite. It is a bit lengthy but precise.
"Adds an HLSL file implementing a version of the Unity core LODDitheringTransition function which is compatible with the Shader Graph package."

If you do rename it, here's my proposed rewrite:
"Adds an HLSL file implementing a version of the Unity core LODDitheringTransition function which is compatible with the Shader Graph package."

How does this sound?

nerites replied:
Those look the same, but I have renamed it so that's fine! Thank you :)

nerites replied:
What about "Adds an HLSL file implementing a version of the Unity core LODDitheringTransition function which can be used in a Shader Graph"? It's not the package itself that's the problem, but how shader graphs only take into account things that are connected into the graph.

ValGrimm-U3D replied:
Perfect!

ValGrimm-U3D commented on com.unity.render-pipelines.core/CHANGELOG.md:
```diff
@@ -12,6 +12,7 @@ The version number for this package has increased due to a version update of a r
 
 ### Added
 - Support for the XboxSeries platform has been added.
+- Added SpeedTree8MaterialUpgrader for common SpeedTree8 code to support HD and Universal's ST8 shadergraphs
```
What does "common" mean in this context? I can guess, but it could be helpful to users to be more explicit.

nerites replied:
How about:
"Added SpeedTree8MaterialUpgrader, which provides utilities for upgrading and importing SpeedTree 8 assets to scriptable render pipelines."

Since technically, it can be used in custom render pipelines, as long as people adhere to the same shader property names...

ValGrimm-U3D replied:
Looks good to me!

ValGrimm-U3D commented on com.unity.render-pipelines.high-definition/CHANGELOG.md:
```diff
@@ -28,6 +28,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Added support for XboxSeries platform.
 - Added an additional check in the "check scene for ray tracing" (case 1314963).
 - Added shader graph unit test for IsFrontFace node
+- Added SpeedTree8 shadergraph as default ST8 shader for HDRP
```
ST8 -> SpeedTree 8

ValGrimm-U3D commented on com.unity.render-pipelines.universal/CHANGELOG.md:
```diff
@@ -22,6 +22,9 @@ The version number for this package has increased due to a version update of a r
 - Fixed issue causing missing shaders on DirectX 11 feature level 10 GPUs. [case 1278390](https://issuetracker.unity3d.com/product/unity/issues/guid/1278390/)
 
+### Added
+- Added SpeedTree8 shadergraph for universal. Since it behaves differently from existing ST8 shader, it is not hooked up or set to be default yet.
```
Proposed:
Added a SpeedTree 8 Shader Graph but did not make it accessible to users or set it as the default. Because URP doesn't yet support per-material culling, this Shader Graph does not yet behave in the same way as the existing handwritten SpeedTree 8 shader for URP.

nerites replied:
It is technically accessible for users. When the universal package is included in their project, it also includes the new shader graph. However, importing and upgrading ST8 assets still uses the original handwritten ST8 shader. 

So, how about:
"Added a SpeedTree 8 Shader  Graph but did not set it as the default when importing or upgrading Speed Tree 8 assets. Because URP doesn't yet support per-material culling, this Shader Graph does not yet behave in the same way as the existing handwritten SpeedTree 8 shader for URP."

ValGrimm-U3D replied:
Swap out "when" in the first sentence with "for" and  then this is good.

ValGrimm-U3D commented:
> 
> 
> Pending docs for lodcrossfade looks good on my end

Docs track with the Unity release rather than with the PR. As a result, the docs are not a blocker for this PR. A version of these will be available for 21.2, most likely by first beta.

nerites commented:
@ValGrimm-U3D I incorporated your feedback in this changeset:
https://github.com/Unity-Technologies/Graphics/pull/3861/commits/b15bf84eeedc852e1d62e5188843986e50ebd6c2

Let me know what you think :) Thanks!

ValGrimm-U3D commented:
> 
> 
> @ValGrimm-U3D I incorporated your feedback in this changeset:
> [b15bf84](https://github.com/Unity-Technologies/Graphics/commit/b15bf84eeedc852e1d62e5188843986e50ebd6c2)
> 
> Let me know what you think :) Thanks!

Thank you (: 
I've looked through your responses and added my own, resolving the conversation where relevant.

nerites commented:
> 
> 
> > @ValGrimm-U3D I incorporated your feedback in this changeset:
> > [b15bf84](https://github.com/Unity-Technologies/Graphics/commit/b15bf84eeedc852e1d62e5188843986e50ebd6c2)
> > Let me know what you think :) Thanks!
> 
> Thank you (:
> I've looked through your responses and added my own, resolving the conversation where relevant.

Additional changes here: b22e9fc

ValGrimm-U3D commented:
> 
> 
> > > @ValGrimm-U3D I incorporated your feedback in this changeset:
> > > [b15bf84](https://github.com/Unity-Technologies/Graphics/commit/b15bf84eeedc852e1d62e5188843986e50ebd6c2)
> > > Let me know what you think :) Thanks!
> > 
> > 
> > Thank you (:
> > I've looked through your responses and added my own, resolving the conversation where relevant.
> 
> Additional changes here: [b22e9fc](https://github.com/Unity-Technologies/Graphics/commit/b22e9fc6fff2aeb0fb172cbba9192bbbef16c719)

Looks great! Thank you. 

(In general I would avoid passive voice, but you've already revised that sentence once, and it isn't worth nitpicking that much in a changelog. )

ValGrimm-U3D reviewed (approved):
Thanks for responding to my many questions and suggestions, fine to merge.  👍

JMargevics reviewed (approved):
https://confluence.unity3d.com/pages/viewpage.action?pageId=129564990 Test doc includes coverage of this backport. No specific issues were found, looks good from HDRP side ✅

nerites commented:
Remaining failures are all hybrid failures discussed here, and are not due to my PR:
https://unity.slack.com/archives/CFZERNTS7/p1617956110148300

erikabar reviewed (changes requested):
we do not have QA bandwidth right now, we will test this after feature cutoff.

nerites commented:
> 
> 
> we do not have QA bandwidth right now, we will test this after feature cutoff.

when is that?

GrantLamb-Unity commented:
@erikabar Do you have an ETA on when we could get a URP QA review?

erikabar commented:
status update: I will finish testing by tomorrow noon

erikabar reviewed (approved):
Tested URP SpeedTree8_PBRLit shadergraph, everything works as expected in the Editor and Standalone build. Also tested Wind Zone, Terrain Tree painting, interaction with Spot, Point, Directional lights and shadows - no issues found.
